### PR TITLE
[FLINK-9777] YARN: JM and TM Memory must be specified with Units

### PR DESCRIPTION
**Pull request by yanghua**
## What is the purpose of the change

*This pull request specify unit for JM and TM memory on YARN mode*

## Brief change log

  - *parse the jm and tm with default MB unit*
  - *change related document*

## Verifying this change

This change is already covered by existing tests*.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / **not documented**)

### CODE REVIEW

**Comment by yanghua** [comment]
@GJL please review~

**Comment by yanghua** [comment]
@zentol can you review this?

**Comment by GJL** [comment]
I will take a look later.

File: flink-yarn/src/main/java/org/apache/flink/yarn/cli/FlinkYarnSessionCli.java [hunk 1]
```diff
@@ -386,10 +386,10 @@ private ClusterSpecification createClusterSpecification(Configuration configurat
 		}
 
 		// JobManager Memory
-		final int jobManagerMemoryMB = MemorySize.parse(configuration.getString(JobManagerOptions.JOB_MANAGER_HEAP_MEMORY)).getMebiBytes();
+		final int jobManagerMemoryMB = MemorySize.parse(configuration.getString(JobManagerOptions.JOB_MANAGER_HEAP_MEMORY), MemorySize.MemoryUnit.MEGA_BYTES).getMebiBytes();
```
Comment (GJL) [inline flink-yarn/src/main/java/org/apache/flink/yarn/cli/FlinkYarnSessionCli.java]: Was there a reason against making `JOB_MANAGER_HEAP_MEMORY_MB` a deprecated key for `JOB_MANAGER_HEAP_MEMORY` when you worked on FLINK-6469? That is:
```
public static final ConfigOption<String> JOB_MANAGER_HEAP_MEMORY =
		key("jobmanager.heap.size")
			.defaultValue("1024m")
			.withDeprecatedKeys("jobmanager.heap.mb")
		    ...
```
I am not sure what happens if someone only configures `jobmanager.heap.mb` in `flink-conf.yaml`. Will the value be respected for all deployment modes?

Same holds for the TaskManager config key.

File: docs/ops/deployment/yarn_setup.md [hunk 2]
```diff
@@ -101,12 +101,12 @@ Usage:
    Optional
      -D <arg>                        Dynamic properties
      -d,--detached                   Start detached
-     -jm,--jobManagerMemory <arg>    Memory for JobManager Container [in MB]
+     -jm,--jobManagerMemory <arg>    Memory for JobManager Container [with unit, if not, use MB]
```
Comment (GJL) [inline docs/ops/deployment/yarn_setup.md]: nit: I would reword this to: 
> *Memory for JobManager container with optional unit (default: MB).*

**Comment by GJL** [comment]
cc: @dawidwys @StephanEwen

File: flink-yarn/src/main/java/org/apache/flink/yarn/cli/FlinkYarnSessionCli.java (hunk 1 above)
Comment (yanghua) [reply flink-yarn/src/main/java/org/apache/flink/yarn/cli/FlinkYarnSessionCli.java]: in FLINK-6469, in order to config the jm's memory with unit I introduced a new key and deprecated `jobmanager.heap.mb`.

* in flink codebase(except shell script) I have removed all the place used `JOB_MANAGER_HEAP_MEMORY_MB` and `jobmanager.heap.mb`, so it will not be used.
* in shell (`config.sh`) the old key `jobmanager.heap.mb` also be supported if the new key `jobmanager.heap.size` can not be read, so it still be supported.

File: docs/ops/deployment/yarn_setup.md (hunk 2 above)
Comment (yanghua) [reply docs/ops/deployment/yarn_setup.md]: change this soon

File: flink-yarn/src/main/java/org/apache/flink/yarn/cli/FlinkYarnSessionCli.java (hunk 1 above)
Comment (GJL) [reply flink-yarn/src/main/java/org/apache/flink/yarn/cli/FlinkYarnSessionCli.java]: Would it hurt to add `.withDeprecatedKeys("jobmanager.heap.mb")`? In case someone does not use the scripts to start the cluster.

Comment (dawidwys) [reply flink-yarn/src/main/java/org/apache/flink/yarn/cli/FlinkYarnSessionCli.java]: I think the problem here would be that we would need to get back to MB as a default unit for this option. Or am I wrong here? The whole idea for introducing the new option was to switch to bytes as a default one, if none provided.

Comment (GJL) [reply flink-yarn/src/main/java/org/apache/flink/yarn/cli/FlinkYarnSessionCli.java]: @dawidwys As I understand, backwards compatibility should be prioritized. See https://github.com/apache/flink/pull/5448#issuecomment-394296226

> We need to add an additional MemoryUnit.parse() method that takes the "default" unit, so that we parse the old heap sizes such that they are in MB if nothing else is specified.

Comment (dawidwys) [reply flink-yarn/src/main/java/org/apache/flink/yarn/cli/FlinkYarnSessionCli.java]: Agreed, missed it previously. Unfortunately `withDeprecatedKeys` won't solve it though. Either we should revert the option (and use the mb versions), or anywhere we use the new option, we should also check for the old one.

Comment (GJL) [reply flink-yarn/src/main/java/org/apache/flink/yarn/cli/FlinkYarnSessionCli.java]: @dawidwys Why is it not enough?

> When obtaining a value from the configuration via Configuration.getValue(ConfigOption), the deprecated keys will be checked in the order provided to this method. The first key for which a value is found will be used - that value will be returned.

Comment (dawidwys) [reply flink-yarn/src/main/java/org/apache/flink/yarn/cli/FlinkYarnSessionCli.java]: Imagine a case when previously we had in configuration:

    jobmanager.heap.mb: 1024

which meant 1024 MB, in the new code we will pass 1024 as value for `jobmanager.heap.size`, which was introduced solely to treat numbers without units as bytes. Therefore it will be parsed as 1024 bytes rather than 1024 MB.

If we switch the default unit to MB, there is no point for having a new `ConfigOption`.

Comment (yanghua) [reply flink-yarn/src/main/java/org/apache/flink/yarn/cli/FlinkYarnSessionCli.java]: to @GJL we can not use `.withDeprecatedKeys("jobmanager.heap.mb")` because `jobmanager.heap.size` and `jobmanager.heap.mb` has different meaning. The former can use different unit such **1g** but the latter can just measure with **MB**.

to @dawidwys and @GJL , now the `jobmanager.heap.mb` just used in config file, and can be calculated accurately, this is used for backwards compatibility, but in the project, it is useless, all the place can be replaced with `jobmanager.heap.size`, and the key in the code could not been exposed to the user?

Comment (yanghua) [reply flink-yarn/src/main/java/org/apache/flink/yarn/cli/FlinkYarnSessionCli.java]: introduce a new config key is used to make the `jobmanager.heap.mb` backwards compatibility in flink config file (config.sh can calculate it accurately). And user can specify the unit for the value of the key `jobmanager.heap.size` .

So if we remove anything about `JOB_MANAGER_HEAP_MEMORY_MB ` in Java and Scala code, is there any problem?

Comment (dawidwys) [reply flink-yarn/src/main/java/org/apache/flink/yarn/cli/FlinkYarnSessionCli.java]: I think we cannot do this as users might not use the `config.sh`. At the same time we think we really should keep backwards compatibility, as otherwise there might be complaints.  We've discussed this with @GJL offline and we think the way to go right now would be to check everywhere we use the new option to check for the old one as well if the new one was not set.

That is:
* `org.apache.flink.client.deployment.ClusterSpecification#fromConfiguration`
* `org.apache.flink.yarn.YarnResourceManager#YarnResourceManager`
* `org.apache.flink.yarn.cli.FlinkYarnSessionCli#createClusterSpecification`

Probably we should wrap it in some utility method.

Also for the yarn command line we should keep the old behaviour. This means we should add "m" suffix in `org.apache.flink.yarn.cli.FlinkYarnSessionCli#applyCommandLineOptionsToConfiguration`

File: flink-core/src/main/java/org/apache/flink/configuration/ConfigurationUtils.java [hunk 3]
```diff
@@ -29,6 +29,46 @@
...
+	 * @param configuration the configuration object
+	 * @return the memory size of job manager's heap memory.
+	 */
+	public static MemorySize getJobManagerHeapMemory(Configuration configuration) {
+		if (configuration.containsKey(JobManagerOptions.JOB_MANAGER_HEAP_MEMORY.key())) {
+			return MemorySize.parse(configuration.getString(JobManagerOptions.JOB_MANAGER_HEAP_MEMORY));
+		} else if (configuration.containsKey(JobManagerOptions.JOB_MANAGER_HEAP_MEMORY_MB.key())) {
+			return MemorySize.parse(configuration.getInteger(JobManagerOptions.JOB_MANAGER_HEAP_MEMORY_MB) + "m");
+		} else {
+			throw new RuntimeException("Can not find config key : " + JobManagerOptions.JOB_MANAGER_HEAP_MEMORY.key()
```
Comment (dawidwys) [inline flink-core/src/main/java/org/apache/flink/configuration/ConfigurationUtils.java]: Use `FlinkRuntimeException`.

File: flink-core/src/main/java/org/apache/flink/configuration/ConfigurationUtils.java [hunk 4]
```diff
@@ -29,6 +29,46 @@
...
+	 * @param configuration the configuration object
+	 * @return the memory size of task manager's heap memory.
+	 */
+	public static MemorySize getTaskManagerHeapMemory(Configuration configuration) {
+		if (configuration.containsKey(TaskManagerOptions.TASK_MANAGER_HEAP_MEMORY.key())) {
+			return MemorySize.parse(configuration.getString(TaskManagerOptions.TASK_MANAGER_HEAP_MEMORY));
+		} else if (configuration.containsKey(TaskManagerOptions.TASK_MANAGER_HEAP_MEMORY_MB.key())) {
+			return MemorySize.parse(configuration.getInteger(TaskManagerOptions.TASK_MANAGER_HEAP_MEMORY_MB) + "m");
+		} else {
+			throw new RuntimeException("Can not find config key : " + TaskManagerOptions.TASK_MANAGER_HEAP_MEMORY.key()
```
Comment (dawidwys) [inline flink-core/src/main/java/org/apache/flink/configuration/ConfigurationUtils.java]: Use `FlinkRuntimeException`.

File: flink-yarn/src/main/java/org/apache/flink/yarn/cli/FlinkYarnSessionCli.java [hunk 5]
```diff
@@ -500,11 +501,11 @@ protected Configuration applyCommandLineOptionsToConfiguration(CommandLine comma
 		}
 
 		if (commandLine.hasOption(jmMemory.getOpt())) {
-			effectiveConfiguration.setString(JobManagerOptions.JOB_MANAGER_HEAP_MEMORY, commandLine.getOptionValue(jmMemory.getOpt()));
+			effectiveConfiguration.setString(JobManagerOptions.JOB_MANAGER_HEAP_MEMORY, commandLine.getOptionValue(jmMemory.getOpt()) + "m");
```
Comment (dawidwys) [inline flink-yarn/src/main/java/org/apache/flink/yarn/cli/FlinkYarnSessionCli.java]: I think we have to be a bit smarter here. The unit might be already provided from cli, right?

File: flink-yarn/src/main/java/org/apache/flink/yarn/cli/FlinkYarnSessionCli.java [hunk 6]
```diff
@@ -500,11 +501,11 @@ protected Configuration applyCommandLineOptionsToConfiguration(CommandLine comma
...
 		}
 
 		if (commandLine.hasOption(tmMemory.getOpt())) {
-			effectiveConfiguration.setString(TaskManagerOptions.TASK_MANAGER_HEAP_MEMORY, commandLine.getOptionValue(tmMemory.getOpt()));
+			effectiveConfiguration.setString(TaskManagerOptions.TASK_MANAGER_HEAP_MEMORY, commandLine.getOptionValue(tmMemory.getOpt()) + "m");
```
Comment (dawidwys) [inline flink-yarn/src/main/java/org/apache/flink/yarn/cli/FlinkYarnSessionCli.java]: Same as above.

File: flink-yarn/src/main/java/org/apache/flink/yarn/cli/FlinkYarnSessionCli.java [hunk 7]
```diff
@@ -500,11 +500,19 @@ protected Configuration applyCommandLineOptionsToConfiguration(CommandLine comma
 		}
 
 		if (commandLine.hasOption(jmMemory.getOpt())) {
-			effectiveConfiguration.setString(JobManagerOptions.JOB_MANAGER_HEAP_MEMORY, commandLine.getOptionValue(jmMemory.getOpt()));
+			String jmMemoryVal = commandLine.getOptionValue(jmMemory.getOpt());
+			if (!jmMemoryVal.toLowerCase().contains("m")) {
```
Comment (dawidwys) [inline flink-yarn/src/main/java/org/apache/flink/yarn/cli/FlinkYarnSessionCli.java]: This won't cover other units than "m". How about checking it with `MemoryUnit.hasUnit`?

File: flink-yarn/src/main/java/org/apache/flink/yarn/cli/FlinkYarnSessionCli.java [hunk 8]
```diff
@@ -500,11 +500,19 @@ protected Configuration applyCommandLineOptionsToConfiguration(CommandLine comma
...
+				jmMemoryVal += "m";
+			}
+			effectiveConfiguration.setString(JobManagerOptions.JOB_MANAGER_HEAP_MEMORY, jmMemoryVal);
 		}
 
 		if (commandLine.hasOption(tmMemory.getOpt())) {
-			effectiveConfiguration.setString(TaskManagerOptions.TASK_MANAGER_HEAP_MEMORY, commandLine.getOptionValue(tmMemory.getOpt()));
+			String tmMemoryVal = commandLine.getOptionValue(tmMemory.getOpt());
+			if (!tmMemoryVal.toLowerCase().contains("m")) {
```
Comment (dawidwys) [inline flink-yarn/src/main/java/org/apache/flink/yarn/cli/FlinkYarnSessionCli.java]: Same as above

**Comment by dawidwys** [comment]
Changes look good now. I think we are still missing tests for the added behaviour. We should check:

- if yarn cli parameters are properly parsed without units
- if yarn cli parameters are properly parsed with units other than m
- if the logic for `getJobManagerHeapMemory/getTaskManagerHeapMemory` works properly (if the old parameter is parsed if the new one is not provided. If the default value is provided if no option was set.

**Comment by yanghua** [comment]
@dawidwys added test case, please review~

File: flink-yarn/src/test/java/org/apache/flink/yarn/FlinkYarnSessionCliTest.java [hunk 9]
```diff
@@ -352,6 +352,76 @@ public void testConfigurationClusterSpecification() throws Exception {
...
+			"yarn");
+
+		CommandLine commandLine = flinkYarnSessionCli.parseCommandLineOptions(args, false);
+
+		ClusterSpecification clusterSpecification = flinkYarnSessionCli.getClusterSpecification(commandLine);
+
+		assertThat(clusterSpecification.getMasterMemoryMB(), is(1024));
+		assertThat(clusterSpecification.getTaskManagerMemoryMB(), is(2048));
+
+		//with unit "m"
```
Comment (dawidwys) [inline flink-yarn/src/test/java/org/apache/flink/yarn/FlinkYarnSessionCliTest.java]: This should be a separate test

File: flink-yarn/src/test/java/org/apache/flink/yarn/FlinkYarnSessionCliTest.java [hunk 10]
```diff
@@ -352,6 +352,76 @@ public void testConfigurationClusterSpecification() throws Exception {
...
+		args = new String[] { "-yn", "2", "-yjm", "1024m", "-ytm", "2048m" };
+		commandLine = flinkYarnSessionCli.parseCommandLineOptions(args, false);
+		clusterSpecification = flinkYarnSessionCli.getClusterSpecification(commandLine);
+
+		assertThat(clusterSpecification.getMasterMemoryMB(), is(1024));
+		assertThat(clusterSpecification.getTaskManagerMemoryMB(), is(2048));
+
+		//with unit non "m"
```
Comment (dawidwys) [inline flink-yarn/src/test/java/org/apache/flink/yarn/FlinkYarnSessionCliTest.java]: This should be a separate test

File: flink-yarn/src/test/java/org/apache/flink/yarn/FlinkYarnSessionCliTest.java [hunk 11]
```diff
@@ -352,6 +352,76 @@ public void testConfigurationClusterSpecification() throws Exception {
...
+		args = new String[] { "-yn", "2", "-yjm", "1g", "-ytm", "2g" };
+		commandLine = flinkYarnSessionCli.parseCommandLineOptions(args, false);
+		clusterSpecification = flinkYarnSessionCli.getClusterSpecification(commandLine);
+
+		assertThat(clusterSpecification.getMasterMemoryMB(), is(1024));
+		assertThat(clusterSpecification.getTaskManagerMemoryMB(), is(2048));
+
+		//specify old config key
```
Comment (dawidwys) [inline flink-yarn/src/test/java/org/apache/flink/yarn/FlinkYarnSessionCliTest.java]: This should be a separate test

File: flink-yarn/src/test/java/org/apache/flink/yarn/FlinkYarnSessionCliTest.java [hunk 12]
```diff
@@ -352,6 +352,76 @@ public void testConfigurationClusterSpecification() throws Exception {
...
+			"yarn");
+
+		commandLine = flinkYarnSessionCli.parseCommandLineOptions(new String[0], false);
+
+		clusterSpecification = flinkYarnSessionCli.getClusterSpecification(commandLine);
+
+		assertThat(clusterSpecification.getMasterMemoryMB(), is(2048));
+		assertThat(clusterSpecification.getTaskManagerMemoryMB(), is(4096));
+
+		//set nothing use default value
```
Comment (dawidwys) [inline flink-yarn/src/test/java/org/apache/flink/yarn/FlinkYarnSessionCliTest.java]: This should be a separate test

**Comment by dawidwys** [comment]
LGTM for me, Thanks for your contribution @yanghua.

@GJL would you like to have last look, or can I merge it?

**Comment by GJL** [comment]
@dawidwys Please merge.
